src/Compilers/CSharp/Portable/Lowering/StateMachineRewriter/IteratorAndAsyncCaptureWalker.cs

```diff
@@ -75,10 +75,17 @@ public static OrderedSet<Symbol> Analyze(CSharpCompilation compilation, MethodSy
                     var variable = kvp.Key;
                     var type = (variable.Kind == SymbolKind.Local) ? ((LocalSymbol)variable).Type : ((ParameterSymbol)variable).Type;
 
-                    foreach (CSharpSyntaxNode syntax in kvp.Value)
+                    if (variable is SynthesizedLocal local && local.SynthesizedKind == SynthesizedLocalKind.Spill)
                     {
-                        // CS4013: Instance of type '{0}' cannot be used inside an anonymous function, query expression, iterator block or async method
-                        diagnostics.Add(ErrorCode.ERR_SpecialByRefInLambda, syntax.Location, type);
+                        diagnostics.Add(ErrorCode.ERR_ByRefTypeAndAwait, local.Locations[0], local.Type);
```

Comment:
>ERR_ByRefTypeAndAwait [](start = 50, length = 21)

Does this correspond to the restricted type case previously in `AwaitExpressionSpiller.AddLocal`? If so, perhaps add `Debug.Assert(local.Type.IsRestrictedType());`.

Reply:
That is true by construction. There is only one line that adds to the set from which the symbol was fetched:

``` c#
            if (type.IsRestrictedType())
            {
                // ...

                _lazyDisallowedCaptures.Add(variable, syntax);
            }
```


---
In reply to: [177570217](https://github.com/dotnet/roslyn/pull/25697#discussion_r177570217) [](ancestors = 177570217)